### ISSUE TITLE
JavaScript Question Summary

### ISSUE THREAD

**Issue by zenz34** [zenz34/FE_Questions#6]
# What is the difference between "var" and "let" keyword?

1. "var" is in JS from the beginning and "let" was introduced in ES6. So "let" won't work in the old browsers which didn't support ES6.

2. "var" has function scope and "let" has block scope.
"var" doesn't respect the block.

3. "var" gets hoisted.

# What's the difference between "==" and "==="?
"==" compares value only.
"===" compares value and type.

# What's the difference between "undefined" and "null"?
1) They both represent empty value.
 "undefined" means you define a variable but not assign a value to it, it's value will automatically be "undefined". Often represents it's not be used yet.
"null" is you are doing it yourself. Like there is a value I wanna clean up, I would set it to "null". Don't set it to be "undefined". 

2) typeof undefined === undefined, typeof null === object
 
# What's the difference between "const" and "let"?
Once "const" variable is assigned a value, it can not be modified.

# how to check if a variable is a null?
if (variable === null && typeof variable === 'object')

# how to check if a variable is a object?
if (typeof x === 'object' && x !== null)

# How to check if a variable's value equals to NaN?
if(!parseFloat(variable) && variable != 0 && typeof variable === "number")
parseFloat returns NaN if a number can not be converted.

#  How to check if a variable's type is Array?
function isArrayFn(o) {
    if (Array.isArray) {
        return Array.isArray(o);
    }

    return (
        Object.prototype.toString.call(o) === Object.prototype.toString.call([]);
    );
}
Object.prototype.toString.call can only return those type:
Arguments, Array, Boolean, Date, Error, Function, JSON, Math, Number, Object, RegExp, String.

#  What is Arrow function and give me an example
An arrow function expression has a shorter syntax than a function expression and does not have its own this, arguments, super, or new.target. These function expressions are best suited for non-method functions, and they cannot be used as constructors.

const profile = {
    name: "",
    setName: function(name) {
        function printOutName() {
            this.name = name;
        }
        printOutName();
    }
};

profile.setName("Jack");
console.log(profile.name === ""); // true
console.log(global.name); // Jack

const profile = {
    name: "",
    setName: function(name) {
        let printOutName = n => {
            this.name = name;
        };
        printOutName(name);
    }
};
profile.setName("Jack");
console.log(profile.name); // Jack

# What is a closure? Can you give a useful example of one?
A closure is a function defined inside another function and has access to its lexical scope even when it is executing outside its lexical scope. The closure has access to variables in three scopes:

Variables declared in its own scope
Variables declared in the scope of the parent function
Variables declared in the global scope
In JavaScript, all functions are closures because they have access to the outer scope, but most functions don't utilize the usefulness of closures: the persistence of state. Closures are also sometimes called stateful functions because of this.

In addition, closures are the only way to store private data that can't be accessed from the outside in JavaScript. They are the key to the UMD (Universal Module Definition) pattern, which is frequently used in libraries that only expose a public API but keep the implementation details private, preventing name collisions with other libraries or the user's own code.

# What is a self-invoked function (IFFE)? What is the usage?
Using parentheses to wrap the entire contents of a JavaScript function, then invoked it using another pair of parentheses.
This technique is very common in JavaScript libraries. It creates a closure around the entire contents of the file which creates a private namespace and thereby helps avoid potential name clashes between different JavaScript modules and libraries. The function is immediately invoked so that the namespace (library name) is assigned the return value of the function.

const myLibrary = (function() {
  var privateVariable = 2
  return {
    publicMethod: () => privateVariable
  }
})()
privateVariable // ReferenceError
myLibrary.publicMethod() // 2

# What is the falsy values?
false
undefined
null
"" (empty string),
NaN
0

# How does prototypal inheritance differ from classical inheritance?
In the classical inheritance paradigm, object instances inherit their properties and functions from a class, which acts as a blueprint for the object. Object instances are typically created using a constructor and the new keyword.

In the prototypal inheritance paradigm, object instances inherit directly from other objects and are typically created using factory functions or Object.create().

# what are the different ways to create an object?
Object literal
Often used to store one occurrence of data.

const person = {
  name: "John",
  age: 50,
  birthday() {
    this.age++
  }
}
person.birthday() // person.age === 51
Constructor
Often used when you need to create multiple instances of an object, each with their own data that other instances of the class cannot affect. The new operator must be used before invoking the constructor or the global object will be mutated.

function Person(name, age) {
  this.name = name
  this.age = age
}
Person.prototype.birthday = function() {
  this.age++
}
const person1 = new Person("John", 50)
const person2 = new Person("Sally", 20)
person1.birthday() // person1.age === 51
person2.birthday() // person2.age === 21
Factory function
Creates a new object similar to a constructor, but can store private data using a closure. There is also no need to use new before invoking the function or the this keyword. Factory functions usually discard the idea of prototypes and keep all properties and methods as own properties of the object.

const createPerson = (name, age) => {
  const birthday = () => person.age++
  const person = { name, age, birthday }
  return person
}
const person = createPerson("John", 50)
person.birthday() // person.age === 51
Object.create()
Sets the prototype of the newly created object.

const personProto = {
  birthday() {
    this.age++
  }
}
const person = Object.create(personProto)
person.age = 50
person.birthday() // person.age === 51
A second argument can also be supplied to Object.create() which acts as a descriptor for the new properties to be defined.

Object.create(personProto, {
  age: {
    value: 50,
    writable: true,
    enumerable: true
  }
})

# What is prototype inheritance?
Basically every object has a property called prototype, we can add properties and methods to it.
When you create a new object from this object, the newly created object will automatcially inherit the properties and methods in the parent object. But not by including these properties, methods in its own properties. Instead, it uses its parent object prototype. The way it works is when you call a perticular property or method, first, it will look its own properties to see if its there. If not there, prototype object.

**Comment by zenz34** [comment]
import store from './configureStore';

describe('Test for store', () => {
  it('Should have the init value', () => {
    const initState = {todos: [], filter: 'all'};
    expect(store.getState()).toEqual(initState);
  });
  it('handle ADD_TODO action', () => {
    const action = {type: 'ADD_TODO', id: 1, text: 'first item'};
    const resultState = {
      todos: [{id: 1, text: 'first item', completed: false}],
      filter: 'all',
    };
    store.dispatch(action);
    expect(store.getState()).toEqual(resultState);
  });
  it('handle TOGGLE action', () => {
    const action = {type: 'TOGGLE_TODO', id: 1};
    const resultState = {
      todos: [{id: 1, text: 'first item', completed: true}],
      filter: 'all',
    };
    store.dispatch(action);
    expect(store.getState()).toEqual(resultState);
  });
  it('handle SET_FILTER action', () => {
    const action = {type: 'SET_FILTER', filter: 'active'};
    const resultState = {
      todos: [{id: 1, text: 'first item', completed: true}],
      filter: 'active',
    };
    store.dispatch(action);
    expect(store.getState()).toEqual(resultState);
  });
  it('stays the same for other unhandled actions', () => {
    const currentState = store.getState();
    const action = {type: 'RANDOM_ACTION'};
    store.dispatch(action);
    expect(store.getState()).toEqual(currentState);
  });
});
